[visionOS] Payment sheet not showing up when a view service tries to present it on a host application owned scene

aprotyas opened the pull request:
#### d9d8f6844b91cd29bf8f8b0182f08a53364ae0a9
<pre>
[visionOS] Payment sheet not showing up when a view service tries to present it on a host application owned scene
<a href="https://bugs.webkit.org/show_bug.cgi?id=268400">https://bugs.webkit.org/show_bug.cgi?id=268400</a>
<a href="https://rdar.apple.com/121673521">rdar://121673521</a>

Reviewed by Tim Horton.

PKPaymentAuthorizationController needs to be vended a scene ID and bundle
ID pair that correspond to each other for it to successfully present a
payment sheet on a scene. Unfortunately, when there is an embedded view
service in a host application, we:

1. Supply a bogus scene ID because view services don&apos;t have an associated
scene ID.
2. Supply the bundle identifier of the view service and not the hosting
application, which actually owns the scene to be presented in.

To address (1), we consult `-[WKUIDelegatePrivate _hostSceneBundleIdentifierForWebView:]`
when we fetch a scene identifier for payment presentation. This ensures
that we supply the _host_ scene identifier to PKPaymentAuthorizationController,
where it can actually present a payment sheet.

However, since we don&apos;t provide the bundle identifier of the host
application, we&apos;re still supplying a mismatching scene/bundle ID pair, so
the controller can&apos;t present a payment sheet successfully. This is
resolved by addressing (2), for which we introduce a new
WKUIDelegatePrivate method `_hostSceneBundleIdentifierForWebView:`. We
consult this when we fetch a bundle identifier for payment presentation,
ensuring that we supply the bundle identifier corresponding to the _host_
application that owns the scene where we would like to present a payment
sheet.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/ios/NetworkConnectionToWebProcessIOS.mm:
(WebKit::NetworkConnectionToWebProcess::getWindowSceneAndBundleIdentifierForPaymentPresentation):
(WebKit::NetworkConnectionToWebProcess::getWindowSceneIdentifierForPaymentPresentation): Deleted.
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.h:
(WebKit::PaymentAuthorizationPresenter::bundleIdentifier const):
* Source/WebKit/Platform/cocoa/PaymentAuthorizationViewController.h:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationViewController.mm:
(WebKit::PaymentAuthorizationViewController::presentInScene):
* Source/WebKit/Platform/ios/PaymentAuthorizationController.h:
* Source/WebKit/Platform/ios/PaymentAuthorizationController.mm:
(-[WKPaymentAuthorizationControllerDelegate presentationSceneBundleIdentifierForPaymentAuthorizationController:]):
(WebKit::PaymentAuthorizationController::presentInScene):
* Source/WebKit/Shared/ApplePay/WebPaymentCoordinatorProxy.h:
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getWindowSceneIdentifierForPaymentPresentation): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
(WebKit::NetworkProcessProxy::getWindowSceneAndBundleIdentifierForPaymentPresentation):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::Internals::getWindowSceneAndBundleIdentifierForPaymentPresentation):
(WebKit::WebPageProxy::Internals::getWindowSceneIdentifierForPaymentPresentation): Deleted.

Canonical link: <a href="https://commits.webkit.org/273801@main">https://commits.webkit.org/273801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eaa2cbfb36a9607e8a400d4e043a37d4e1a9416b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39301 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32845 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18057 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31449 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32405 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11500 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11511 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40547 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33202 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33012 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37439 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11774 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9612 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35555 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13450 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8319 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12185 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12660 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eddydas reviewed (approved):
LGTM

hortont424 commented on Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
```diff
@@ -123,4 +128,27 @@
 #endif
 
+#if ENABLE(APPLE_PAY_REMOTE_UI_USES_SCENE)
+void NetworkProcessProxy::getWindowSceneAndBundleIdentifierForPaymentPresentation(WebPageProxyIdentifier webPageProxyIdentifier, CompletionHandler<void(const String&, const String&)>&& completionHandler)
```
"get" is usually reserved for functions with out args, which this is not

hortont424 replied:
But, this is preexisting, so maybe?

aprotyas replied:
It's a pretty small API surface, so I'll just change it to `windowSceneAndBundleIdentifierForPaymentPresentation`

aprotyas replied:
Well... actually, that makes this call read very awkwardly? I'll just keep it as-is for now.

```cpp
    m_client.windowSceneAndBundleIdentifierForPaymentPresentation(webPageProxyID, [weakThis = WeakPtr { *this }, completionHandler = WTFMove(completionHandler)](const String& sceneIdentifier, const String& bundleIdentifier) mutable {
        if (!weakThis) {
            completionHandler(false);
            return;
        }

        if (!weakThis->m_authorizationPresenter) {
            completionHandler(false);
            return;
        }

        weakThis->m_authorizationPresenter->presentInScene(sceneIdentifier, bundleIdentifier, WTFMove(completionHandler));
    });
```

cdumez replied:
Please convert the `weakThis` into a `RefPtr` right away, before you use it:
```
RefPtr protectedThis = weakThis.get();
if (!protectedThis) {
    completionHandler(false);
    return;
}
...
```

aprotyas replied:
Thanks for catching that. I'll fix it before landing. I should've read through the code a bit more carefully 🙂

EDIT: Turns out that WebPaymentCoordinatorProxy isn't RefCounted. It arguably should be given it does async work, but that's out of scope of this PR.

aprotyas committed:
[visionOS] Payment sheet not showing up when a view service tries to present it on a host application owned scene
https://bugs.webkit.org/show_bug.cgi?id=268400
rdar://121673521

Reviewed by Tim Horton.

PKPaymentAuthorizationController needs to be vended a scene ID and bundle
ID pair that correspond to each other for it to successfully present a
payment sheet on a scene. Unfortunately, when there is an embedded view
service in a host application, we:

1. Supply a bogus scene ID because view services don't have an associated
scene ID.
2. Supply the bundle identifier of the view service and not the hosting
application, which actually owns the scene to be presented in.

To address (1), we consult `-[WKUIDelegatePrivate _hostSceneBundleIdentifierForWebView:]`
when we fetch a scene identifier for payment presentation. This ensures
that we supply the _host_ scene identifier to PKPaymentAuthorizationController,
where it can actually present a payment sheet.

However, since we don't provide the bundle identifier of the host
application, we're still supplying a mismatching scene/bundle ID pair, so
the controller can't present a payment sheet successfully. This is
resolved by addressing (2), for which we introduce a new
WKUIDelegatePrivate method `_hostSceneBundleIdentifierForWebView:`. We
consult this when we fetch a bundle identifier for payment presentation,
ensuring that we supply the bundle identifier corresponding to the _host_
application that owns the scene where we would like to present a payment
sheet.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/ios/NetworkConnectionToWebProcessIOS.mm:
(WebKit::NetworkConnectionToWebProcess::getWindowSceneAndBundleIdentifierForPaymentPresentation):
(WebKit::NetworkConnectionToWebProcess::getWindowSceneIdentifierForPaymentPresentation): Deleted.
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.h:
(WebKit::PaymentAuthorizationPresenter::bundleIdentifier const):
* Source/WebKit/Platform/cocoa/PaymentAuthorizationViewController.h:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationViewController.mm:
(WebKit::PaymentAuthorizationViewController::presentInScene):
* Source/WebKit/Platform/ios/PaymentAuthorizationController.h:
* Source/WebKit/Platform/ios/PaymentAuthorizationController.mm:
(-[WKPaymentAuthorizationControllerDelegate presentationSceneBundleIdentifierForPaymentAuthorizationController:]):
(WebKit::PaymentAuthorizationController::presentInScene):
* Source/WebKit/Shared/ApplePay/WebPaymentCoordinatorProxy.h:
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getWindowSceneIdentifierForPaymentPresentation): Deleted.
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
(WebKit::NetworkProcessProxy::getWindowSceneAndBundleIdentifierForPaymentPresentation):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::Internals::getWindowSceneAndBundleIdentifierForPaymentPresentation):
(WebKit::WebPageProxy::Internals::getWindowSceneIdentifierForPaymentPresentation): Deleted.

Canonical link: https://commits.webkit.org/273801@main
